File: dsa-interview-crash-course/1004_max_consecutive_ones_iii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function (nums, k) {
  let l = 0,
    max_length = 0;

  for (let i = 0; i < nums.length; ++i) {
    if (nums[i] === 0) {
      --k;
    }

    if (k < 0) {
      if (nums[l] === 0) {
        ++k;
      }
      ++l;
    }

    max_length = Math.max(max_length, i - l + 1);
  }
  return max_length;
};

File: leetcode-75/1004_max_consecutive_ones_III.py
from typing import List


class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        i = 0
        j = 0
        nZeros = 0
        maxOnes = 0
        while j < len(nums):
            if nums[j] == 0:
                nZeros += 1
            if nZeros > k:
                while nums[i] == 1:
                    i += 1
                i += 1
                nZeros = k
            maxOnes = max(maxOnes, j - i + 1)
            j += 1

        return maxOnes